book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+In this section, we're going to see how TypeScript can help when a value is one of many possible types. We'll first look at declaring those types using union types, then we'll see how TypeScript can narrow down the type of a value based on your runtime code.
+
+## Unions and Literals
+
+### Union Types
+
+A union type is TypeScript's way of saying that a value can be "either this type or that type".
+
+This situation comes up in JavaScript all the time. Imagine you have a value that is a `string` on Tuesdays, but `null` the rest of the time:
+
+```ts twoslash
+const message = Date.now() % 2 === 0 ? "Hello Tuesdays!" : null;
+//    ^?
+```
+
+If we hover over `message`, we can see that TypeScript has inferred its type as `string | null`.
+
+This is a union type. It means that `message` can be either a `string` or `null`.
+
+#### Declaring Union Types
+
+We can declare our own union types.
+
+For example, you might have an `id` that could be either a `string` or a `number`:
+
+```ts twoslash
+const logId = (id: string | number) => {
+  console.log(id);
+};
+```
+
+This means that `logId` can accept either a `string` or a `number` as an argument, but not a `boolean`:
+
+```ts twoslash
+// @errors: 2345
+const logId = (id: string | number) => {
+  console.log(id);
+};
+// ---cut---
+logId("abc");
+
+logId(123);
+
+logId(true);
+```
+
+To create a union type, the `|` operator is used to separate the types. Each type of the union is called a 'member' of the union.
+
+Union types also work when creating your own type aliases. For example, we can refactor our earlier definition into a type alias:
+
+```typescript
+type Id = number | string;
+
+function logId(id: Id) {
+  console.log(id);
+}
+```
+
+Union types can contain many different types - they don't all have to be primitives, or don't need to be related in any way. When they get particularly large, you can use this syntax (with the `|` before the first member of the union) to make it more readable:
+
+```typescript
+type AllSortsOfStuff =
+  | string
+  | number
+  | boolean
+  | object
+  | null
+  | {
+      name: string;
+      age: number;
+    };
+```
+
+Union types can be used in many different ways, and they're a powerful tool for creating flexible type definitions.
+
+### Literal Types
```

Comment:
it feels slightly off that this section is about literal types and then examples are focused on unions of literal types, it seems like this takes a small shortcut in introducing concepts. To fix that maybe you could at least rename this header to smth like "Literal types unions" and just add a sentence or two to the content of this paragraph to make the flow slightly more... flowing 😅 Like, "let's first take a look at literal types [...] and now let's see how powerful those are when used in unions"